main/src/main/scala/sbt/Defaults.scala

```diff
@@ -1849,15 +1863,31 @@ object Defaults extends BuildCommon {
     Def.inputTask {
       val handle = bgRunMain.evaluated
       val service = bgJobService.value
-      service.waitForTry(handle).get
+      val log = streams.value.log
+      val scope = resolvedScoped.value
+      val st = state.value
+      val background = runInBackground.value && StandardMain.exchange.hasServer
+      st.get(Keys.taskTerminal) match {
+        case Some(term) if background =>
+          BackgroundRunTask(term, "fg-run-main", log, st, scope)(() => service.waitForTry(handle))
+        case _ => service.waitForTry(handle).get
+      }
     }
 
   // run calls bgRun in the background and waits for the result.
   def foregroundRunTask: Initialize[InputTask[Unit]] =
     Def.inputTask {
       val handle = bgRun.evaluated
       val service = bgJobService.value
-      service.waitForTry(handle).get
+      val log = streams.value.log
+      val scope = resolvedScoped.value
+      val st = state.value
+      val background = runInBackground.value && StandardMain.exchange.hasServer
+      st.get(Keys.taskTerminal) match {
+        case Some(term) if background =>
+          BackgroundRunTask(term, "fg-run", log, st, scope)(() => service.waitForTry(handle))
+        case _ => service.waitForTry(handle).get
+      }
```

Comment:
I wonder if this could be implemented at the command processing or client / `CommandChannel` level. Like if a client sees `run` key, it would actually dispatch `bgRun` task and then block only the channel. This allows remaining channels to funnel commands into the exchange.

Reply:
That is a potential alternative approach. I think that was my first idea but I ruled it out. Most of the logic of this pr is designed to address these two issues:
1. How do you ensure that the ui doesn’t switch to the shell after the bg job stops
2. How do you defer the completion of the exec until the bg job has completed?

If we just did a simple command substitution of run to bgrun, both of those issues crop up.

Reply:
> 1. How do you ensure that the ui doesn’t switch to the shell after the bg job stops

In the original sbt server circa sbt 1.1.0, there was `ConsoleUnpromptEvent` (https://github.com/sbt/sbt/blob/v1.1.0/main-command/src/main/scala/sbt/internal/CommandChannel.scala) and the idea was that if one channel queues a command like `compile`, we would have the ability to block out all the clients as if they have also typed in `compile`. This use case sound similar to that.

> 2. How do you defer the completion of the exec until the bg job has completed?

This I'm not sure. Exec could end after bg job is done. If there's a BSP events attached to it, then those would need to be deferred..

Another more radical idea might be to fork from the client. Traditionally sbt has always emulated `run` as function calls in sandbox ClassLoader, but for the longevity of server process, it might be more stable to only calculate the copied classpath, and let each client do the running however they please.

Reply:
You're on to something with forking from the client though I think it would make more sense to fork from the server. The ideal approach for testing and run is always `fork := true` so that the state of sbt doesn't affect run and test. The problem, of course, is that the performance is much worse in many cases. A workaround would be to create a test/run daemon that has an ipc or socket connection with the server. The daemon would create its own classloader cache and we could send it the classpath layers (probably using a json protocol) so that the performance of running on the daemon would be indistinguishable from running in process with classloader caching. But there also could be logic for restarting the daemon. For example, we could easily detect a memory leak in the daemon and kill it if it using more memory after a few runs. We also could add an idle timeout for the daemon.

The general solution is that sbt should fork a remote process for every task. sbt is kind of backwards in that in-process evaluation is the default but it is actually a special case of the general solution.